scripts/clean-up-localizations.js

```diff
@@ -0,0 +1,75 @@
+const fs = require('fs');
+const glob = require('glob');
+
+const localizationFile_default = 'src/assets/i18n/en_US.json';
+
+// regular expression for patterns of not explicitly used localization keys (dynamic created keys, error keys from REST calls)
+// ADDITIONAL PATTERNS HAVE TO BE ADDED HERE
+const regEx = /account\.login\..*\.message|.*\.error.*/i;
+
+// store localizations from default localization file in an object
+var localizations_default = JSON.parse(fs.readFileSync(localizationFile_default, 'utf8'));
+console.log('Clean up file', localizationFile_default, 'as default localization file');
+
+// add not explicitly used localization keys with their localization values
+var localizationsFound = {};
+Object.keys(localizations_default)
+  .filter(localization => regEx.test(localization))
+  .map(localizationKey => {
+    localizationsFound[localizationKey] = localizations_default[localizationKey];
+    delete localizations_default[localizationKey];
+  });
+
+// go through directory recursively and find files to be searched
+const filesToBeSearched = glob.sync('{src,projects}/**/!(*.spec).{ts,html}');
+
+// add used localization keys with their localization values
+filesToBeSearched.forEach(filePath => {
+  const fileContent = fs.readFileSync(filePath);
+  for (var localizationKey in localizations_default) {
+    if (fileContent.includes(localizationKey)) {
+      // store found localizations
+      localizationsFound[localizationKey] = localizations_default[localizationKey];
+      delete localizations_default[localizationKey];
+    }
+  }
+});
+
+// log removed localization keys
+Object.keys(localizations_default).forEach(localizationKey => {
+  console.log('\x1b[31m%s\x1b[0m', `  Localization key ${localizationKey} removed because it is not used`);
+});
+
+// sort found localizations
+var localizationsFoundOrdered = {};
+Object.keys(localizationsFound)
+  .sort()
+  .forEach(key => {
+    localizationsFoundOrdered[key] = localizationsFound[key];
+  });
+
+// write found localizations into default localization file
+fs.writeFileSync(localizationFile_default, JSON.stringify(localizationsFoundOrdered, null, 2) + '\n');
+
+// find localization files for other languages
+var localizationFiles_lang = [];
+glob.sync('src/assets/i18n/*.json').forEach(filePath => {
+  if (filePath !== localizationFile_default) {
+    localizationFiles_lang.push(filePath);
+  }
+});
+
+// create cleaned localization files for other languages
+localizationFiles_lang.forEach(file => {
+  console.log('\nClean up file', file);
+  var localizations_lang = JSON.parse(fs.readFileSync(file, 'utf8'));
+  // find missing localization keys
+  _.difference(Object.keys(localizationsFoundOrdered), Object.keys(localizations_lang)).forEach(key => {
+    console.log('\x1b[33m%s\x1b[0m', `  Localization key ${key} not found in ${file}`);
+  });
+  fs.writeFileSync(
+    file,
+    JSON.stringify(_.pick(localizations_lang, Object.keys(localizationsFoundOrdered)), null, 2) + '\n'
```

Comment:
instead of adding the newline, it might just be better, to run prettier on the changed files: `execSync('npx prettier --write ' + path);`

Reply:
Yes, this is more or less an act of desperation since we assume - but cannot actually prove in the CI pipeline - that this missing new line at the file ending might be the problem. Since it does not resolve the problem we will remove this hack again. CI runs with localization files commited without a newline at the end did not resolve the problem either.
The idea with an additional prettier run sounds good. I will test that. Or maybe I will move the localization check in front of the formatting run within 'npm run check'? Thanks for the idea.

Reply:
If you change the order in `npm run check`, then the script itself will not produce proper output when run individually. I would prefer the manual run in the script itself. We do this in many scripts, and it's running quite fast when pointed to individual files 😉

Reply:
Added the prettier run to the script. Unfortunately it does not resolve the problem with the diffs in the localization files. :(

Reply:
exchanged `git diff-files` with `git diff`.